Only use localStorage when it is available

vdonato opened the pull request:
We ran into an issue with theming on Streamlit Sharing on Chrome
Incognito mode because streamlit apps are run in an iframe so are
considered third party code.

Chrome Incognito disallows third party cookies/data by default, which
was causing the app to fail to load in this case. A (temporary) solution
is to simply avoid trying using `localStorage` when it's unavailable to
us.

vdonato committed:
Only use localStorage when it is available

We ran into an issue with theming on Streamlit Sharing on Chrome
Incognito mode because streamlit apps are run in an iframe so are
considered third party code.

Chrome Incognito disallows third party cookies/data by default, which
was causing the app to fail to load in this case. A (temporary) solution
is to simply avoid trying using `localStorage` when it's unavailable to
us.

kmcgrady reviewed (approved):
Looks good overall and nice cleanup!

kmcgrady commented on frontend/src/theme/utils.ts:
```diff
@@ -427,20 +427,59 @@ export const getSystemTheme = (): ThemeConfig => {
+// Method taken from
+// https://stackoverflow.com/questions/16427636/check-if-localstorage-is-available
+export const localStorageAvailable = (): boolean => {
+  const { localStorage } = window
```
Looks like the error still exists in incognito. I wonder if just accessing `localStorage` from window should be in the try/catch block.

vdonato replied:
:+1: that was it. Thanks!

vdonato committed:
Remove leftover .only

vdonato committed:
Grab localStorage in try block too

vdonato committed:
Include pathname in key for cached theme

This is necessary as a browser's localStorage is scoped to a domain,
so all apps on Streamlit sharing will have the same key for their cached
theme (so will overwrite each other) if we don't do this.
